File: 567_Permutation_in_String/567_1.c
bool checkInclusion(char * s1, char * s2){
    int len1 = strlen(s1), len2 = strlen(s2);
    int map[26];
    int windowCnt[26];
    memset(map, 0, sizeof(map));
    memset(windowCnt, 0, sizeof(windowCnt));
    for (int i = 0; i < len1; i++) {
        map[s1[i] - 'a']++;
    }
    
    int left = 0, right = 0, distance = 0;
    for (int i = 0; i < 26; i++) {
        if (map[i] != 0) distance++;
    }
    
    while (right < len2) {
        char c = s2[right++] - 'a';
        if (map[c] > 0) {
            windowCnt[c]++;
            if (map[c] == windowCnt[c])
                distance--;
        }

        if (distance == 0) 
            return true;

        while (right - left >= len1) {
            
            char d = s2[left++] - 'a';
            if (map[d] > 0) {
                if (map[d] == windowCnt[d])
                    distance++;
                windowCnt[d]--;
            }
        }
    }
    return false;
}

int main(void) {
    printf("%s\n", checkInclusion("hello", "ooolleoooleh") ? "true" : "false");
}

File: 567_Permutation_in_String/567_1.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        char_count   = [0 for _ in range(26)]
        window_count = [0 for _ in range(26)]
        for c in s1:
            char_count[ord(c) - ord('a')] += 1
        
        left = right = distance = 0
        for i in range(26):
            if char_count[i] > 0:
                distance += 1
        
        while right < len(s2):
            c = ord(s2[right]) - ord('a')
            right += 1
            if char_count[c] > 0:
                window_count[c] += 1
                if char_count[c] == window_count[c]:
                    distance -= 1
    
            if distance == 0:
                return True

            while right - left >= len(s1):
                d = ord(s2[left]) - ord('a')
                left += 1
                if char_count[d] > 0:
                    if char_count[d] == window_count[d]:
                        distance += 1
                    window_count[d] -= 1
        
        return False